presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseToMap.java

```diff
@@ -0,0 +1,278 @@
+            List<RowExpression> thens = new ArrayList<RowExpression>(node.getArguments().size());
+            RowExpression checkExpr;
+            int start;
+
+            if (!(firstArg instanceof SpecialFormExpression && ((SpecialFormExpression) firstArg).getForm().equals(WHEN))) {
+                if (firstArg.equals(constant(true, BOOLEAN))) {
+                    // We generate weird CASE (true) WHEN p1 THEN v1 etc. for non-searched case
+                    // So drop the true
+                    checkExpr = null;
+                }
+                else {
+                    checkExpr = firstArg;
+                }
+
+                start = 1;
+            }
+            else {
+                checkExpr = null;
+                start = 0;
+            }
+
+            for (int i = start; i < numArgs - 1; i++) {
+                RowExpression whenClause = node.getArguments().get(i);
+                RowExpression value = whenClause.getChildren().get(1);
+
+                if (!(value instanceof ConstantExpression)) {
+                    // THEN is not a constant
+                    return node;
+                }
+
+                RowExpression when = whenClause.getChildren().get(0);
+                RowExpression key;
+                RowExpression curCheck;
+
+                if (when instanceof ConstantExpression) {
+                    if (!addKeyValue(when, keySet, whens, value, thens)) {
+                        return node;
+                    }
+                }
+                else if (logicalRowExpressions.isEqualsExpression(when)) {
+                    RowExpression lhs = when.getChildren().get(0);
+                    RowExpression rhs = when.getChildren().get(1);
+
+                    if (!lhs.getType().equals(rhs.getType())) {
+                        // We keep it simple
+                        return node;
+                    }
+
+                    if (lhs instanceof ConstantExpression) {
+                        curCheck = rhs;
+                        key = lhs;
+                    }
+                    else if (rhs instanceof ConstantExpression) {
+                        curCheck = lhs;
+                        key = rhs;
+                    }
+                    else {
+                        return node;
+                    }
+
+                    if (checkExpr == null) {
+                        checkExpr = curCheck;
+                    }
+                    else if (!curCheck.equals(checkExpr)) {
+                        return node;
+                    }
+
+                    if (!addKeyValue(key, keySet, whens, value, thens)) {
+                        return node;
+                    }
+                }
+                else if (when instanceof SpecialFormExpression && ((SpecialFormExpression) when).getForm() == IN) {
+                    curCheck = ((SpecialFormExpression) when).getArguments().get(0);
+                    if (checkExpr == null) {
```

Comment:
this is repeated several times, can we refactor it?

Reply:
Yeah - it was annoying because we still need to check and return i the check fails so no point in writing a separate function